src/Features/Core/Portable/SplitOrMergeIfStatements/AbstractMergeIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,91 @@
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SplitOrMergeIfStatements
+{
+    internal abstract class AbstractMergeIfStatementsCodeRefactoringProvider<TExpressionSyntax> : CodeRefactoringProvider
+        where TExpressionSyntax : SyntaxNode
+    {
+        protected abstract string IfKeywordText { get; }
+
+        protected abstract bool IsApplicableSpan(SyntaxNode node, TextSpan span, out SyntaxNode ifStatementNode);
+
+        protected abstract bool IsIfStatement(SyntaxNode node);
+
+        protected abstract CodeAction CreateCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument);
+
+        protected abstract Task<bool> CanBeMergedAsync(
+            Document document, SyntaxNode ifStatement, ISyntaxFactsService syntaxFacts, CancellationToken cancellationToken);
+
+        protected abstract SyntaxNode GetChangedRoot(
+            SyntaxNode root, SyntaxNode ifStatement, ISyntaxFactsService syntaxFacts, SyntaxGenerator generator);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span, getInnermostNodeForTie: true);
+
+            if (IsApplicableSpan(node, context.Span, out var ifStatement))
+            {
+                var syntaxFacts = context.Document.GetLanguageService<ISyntaxFactsService>();
+
+                if (await CanBeMergedAsync(context.Document, ifStatement, syntaxFacts, context.CancellationToken).ConfigureAwait(false))
+                {
+                    context.RegisterRefactoring(
+                        CreateCodeAction(
+                            c => RefactorAsync(context.Document, context.Span, syntaxFacts, c)));
+                }
+            }
+        }
+
+        private async Task<Document> RefactorAsync(Document document, TextSpan span, ISyntaxFactsService syntaxFacts, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(span, getInnermostNodeForTie: true);
+
+            Contract.ThrowIfFalse(IsApplicableSpan(node, span, out var ifStatement));
+
+            var newRoot = GetChangedRoot(root, ifStatement, syntaxFacts, document.GetLanguageService<SyntaxGenerator>());
```

Comment:
strange (to me) that you had to get the facts early and pass them here, and that you get hte generator, but dont' use it here, but just pass along.  Why not just pass the document along?

Reply:
Sure, but it might be unclear as to why the whole document is needed if it's only needed for language services - which don't require the content of the document at all, the only information they need is the language.

Reply:
➡️ Fixed in dc77a01fe5f1e6fd0f96e94402ea7e9c5b8ee73c